mathics/core/builtin.py

```diff
@@ -923,18 +926,26 @@ def eval_iter(self, expr, i, imin, imax, di, evaluation):
                 return result
             return
 
-        index = imin.evaluate(evaluation)
+        index = Integer0
+        imin = imin.evaluate(evaluation)
         imax = imax.evaluate(evaluation)
         di = di.evaluate(evaluation)
 
+        # (imax - imin) / di
+        normalised_range = Expression(
+            Symbol("System`Chop"),
+            Expression(
+                SymbolTimes,
+                Expression(SymbolPlus, imax, Expression(SymbolTimes, IntegerM1, imin)),
+                Expression(SymbolPower, di, IntegerM1),
+            ),
+        ).evaluate(evaluation)
+
         result = []
-        compare_type = (
-            SymbolGreaterEqual
-            if Expression(SymbolLess, di, Integer0).evaluate(evaluation).to_python()
-            else SymbolLessEqual
-        )
         while True:
-            cont = Expression(compare_type, index, imax).evaluate(evaluation)
+            cont = Expression(SymbolLessEqual, index, normalised_range).evaluate(
```

Comment:
With this change, I would like to add some extra pytests checking what happens when `di=0`, or `di=Infinity`.